sdk_v2/kubeflow/training/api/training_client.py

```diff
@@ -12,7 +12,553 @@
+                        num_procs = item["spec"]["mlPolicy"]["torch"]["numProcPerNode"]
+
+                    # Get the device count per Trainer node.
+                    # TODO (andreyvelich): Currently, we get the device type from
+                    # the runtime labels.
+                    _, device_count = utils.get_container_devices(resources, num_procs)
+                    if device_count != constants.UNKNOWN:
+                        device_count = str(
+                            int(device_count)
+                            * int(item["spec"]["mlPolicy"]["numNodes"])
+                        )
+
+                    runtime = types.Runtime(
+                        name=item["metadata"]["name"],
+                        phase=(
+                            item["metadata"]["labels"][constants.PHASE_KEY]
+                            if constants.PHASE_KEY in item["metadata"]["labels"]
+                            else constants.UNKNOWN
+                        ),
+                        device=(
+                            item["metadata"]["labels"][constants.DEVICE_KEY]
+                            if constants.DEVICE_KEY in item["metadata"]["labels"]
+                            else constants.UNKNOWN
+                        ),
+                        device_count=device_count,
+                    )
+
+                    result.append(runtime)
+        except multiprocessing.TimeoutError:
+            raise TimeoutError(
+                f"Timeout to list {constants.CLUSTER_TRAINING_RUNTIME_KIND}s "
+                f"in namespace: {self.namespace}"
+            )
+        except Exception:
+            raise RuntimeError(
+                f"Failed to list {constants.CLUSTER_TRAINING_RUNTIME_KIND}s "
+                f"in namespace: {self.namespace}"
+            )
+
+        return result
+
+    def train(
+        self,
+        runtime_ref: str,
+        trainer: Optional[types.Trainer] = None,
+        dataset_config: Optional[types.HuggingFaceDatasetConfig] = None,
+        model_config: Optional[types.HuggingFaceModelInputConfig] = None,
+    ) -> str:
+        """Create the TrainJob. TODO (andreyvelich): Add description
+
+        Returns:
+            str: The unique name of the TrainJob that has been generated.
+
+        Raises:
+            ValueError: Input arguments are invalid.
+            TimeoutError: Timeout to create TrainJobs.
+            RuntimeError: Failed to create TrainJobs.
+        """
+
+        if trainer:
+            utils.validate_trainer(trainer)
+
+        # Generate unique name for the TrainJob.
+        # TODO (andreyvelich): Discuss this TrainJob name generation.
+        train_job_name = random.choice(string.ascii_lowercase) + uuid.uuid4().hex[:11]
+
+        # Build the Trainer.
+        trainer_crd = models.KubeflowOrgV2alpha1Trainer()
+
+        # Add number of nodes to the Trainer.
+        if trainer and trainer.num_nodes:
+            trainer_crd.num_nodes = trainer.num_nodes
```

Comment:
What if the only runtime has num_nodes information.

Reply:
In that case, number of nodes from the runtime will be propagated to the JobSet.
E.g. number of nodes in the TrainJob's Trainer will be empty.

Reply:
Ah, that's right. This `trainer` is not CR Trainer, and this is Python trainer API.